### ISSUE TITLE
[Question] Specialize string_traits for binary data block

### ISSUE THREAD

**Issue by bka2004** [jtv/libpqxx#698]
From reading the docs I got the impression that in "to_buf"/"into_buf" I create a string that represents my data.
And in "from_string" I get a string with exactly that representation and I can do a reverse transformation.

But I observed that in from_string, I get a hex-string that represents the bytes that form the string I create in "to_buf"/"into_buf". Is that always so? Or am I doing something wrong?

Another question is if the string created in to_buf/into_buf may contain zero bytes? The result must be zero terminated, but as a pointer to the end is returned and the beginning is also known, perhaps the size is also stored somewhere? I'm specializing this for a data type that holds a binary data block, so zero bytes could be in there.

Currently, I'm just writing any byte from the data block to the chars from the buffer (in to_buf/into_buf). In from_string, I decode the hex-string and get the desired result, but I haven't tested yet with data that contains zero bytes.
I also wonder if there's a way to avoid the hex-string decoding, which takes some time and performance is crucial in our application.

**Comment by jtv** [comment]
Yes, `to_buf()`/`into_buf()`/`to_string()` converts a string of `std::byte` into a hex representation.  And `from_string<std::basic_string<std::byte>>()` should convert that hex representation into a binary string.  Or it converts an `int` to a series of digits, and so on.  More generally, `to_buf()` and friends convert an object of your type to a text format that postgres understands, and `from_string()` converts in the other direction.

On the other hand, `from_string()` _can_ return something that contains zero bytes.  That's because it's entirely your business what that object is.

The string representation that `to_buf()` etc. produce must _not_ contain zero bytes.  (Apart from the trailing zero of course.)  It should be ASCII text.  Bear in mind here that libpqxx interfaces with libpq, where pretty much everything is zero-terminated strings, so there's no option to work in the "pointer-and-size" paradigm.  The caller of `into_buf()` can store that pointer, but inside the implementation for `into_buf()` there's just no place for it to go.

(I may be misunderstanding you or be a bit unclear.  Didn't get much sleep.)

**Comment by jtv** [comment]
@bka2004 I hope that answered your question?  I'll keep the ticket open for now, to make it easy to continue the conversation.

**Comment by bka2004** [comment]
To be honest, I'm still a bit confused. I'll explain why:
In my current situation I have an instance of my own data type representing a binary block of data which starts with the values 0x3c, 0x3f, 0x78.

To avoid problems with the value  0x00, in "into_buf", I no longer directly put these values into the provided char buffer, but I create a hex string from the values and insert the characters from the hex string into the char buffer.
Hex string starts with "3C3F78", so the first char is filled with 0x33, then 0x43, 0x33 and so on.

In "from_string" (when reading back the value from the database) I now expect to get back the same string as I have created in "into_buf": "3C3F78...."

BUT I get back a string like this: "\x334333463738...."
Which is again a hex representation of my hex representation.

My question is: why?? Am I doing something wrong? Why can't I get the same string back as I created? Decoding from hex string representation to bytes twice seems unnecessary.

**Comment by jtv** [comment]
The `\x334333463738....` is indeed the database's text representation of the binary data you put in.

If you want your hex data interpreted as a text encoding of binary data, you should prefix the hex bytes with `\x` as well.  Otherwise, what you're putting in is just some ASCII data that happens to be hexadecimal digits.  (And I would suggest using lower case to match the format coming the other way, though I don't think it matters too much.)  Also of course, ensure that those extra two bytes are included in the size that your `size_buffer()` returns.

If all you really want to store on the database side is a BYTEA, the simplest way to build your conversion is to make the functions in your `string_traits` specialisation call the existing specialisation for `std::basic_string_view<std::byte>`.

**Comment by bka2004** [comment]
Alright, both hints worked flawlessly. Including "\x" in my own string and just forwarding to string_trains<std::basic_string_view<std::byte>> (for from_string I had to use string_traits<std::basic_string<std::byte>> cause there's no from_string for string_view).

Thanks a lot, you can close this issue now.

**Comment by jtv** [comment]
OK.  And I'm considering supporting conversion to `string_view`, under the proviso that it's your own responsibility to keep an eye on the content buffer's actual lifetime.  But it'll mean a lot of documenting.
